cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/installqueue/InstallQueue.scala

```
@@ -0,0 +1,280 @@
+          }
+      }
+    ).forPath(
+      s"$pendingOperationsPath/${pc.as[String]}"
+    )
+    promise
+  }
+
+  private[this] def getPendingCoordinates: Future[List[PackageCoordinate]] = {
+    val promise = Promise[List[PackageCoordinate]]()
+    client.getChildren.inBackground(
+      handler { case event if event.getType == CuratorEventType.CHILDREN =>
+        val code = KeeperException.Code.get(event.getResultCode)
+        code match {
+          case KeeperException.Code.OK =>
+            val pcs = event.getChildren.toList.map(_.as[Try[PackageCoordinate]].toOption.get) //TODO
+            promise.setValue(pcs)
+          case KeeperException.Code.NONODE =>
+            promise.setValue(List())
+          case _ =>
+            promise.setException(KeeperException.create(code, event.getPath))
+        }
+      }
+    ).forPath(
+      pendingOperationsPath
+    )
+    promise
+  }
+
+  /** Signals to the install queue that the operation on the package
+    * failed. This will "move" the package to the error pile.
+    *
+    * @param packageCoordinate The package coordinate of the
+    *                          package on which the operation failed
+    * @param operation         The operation that failed
+    * @param failure           The failure that occurred
+    * @return a future that will complete when state changes has been completed
+    */
+  override def failure(
+    packageCoordinate: PackageCoordinate,
+    operation: Operation,
+    failure: ErrorResponse
+  ): Future[Unit] = {
+    val operationFailure = OperationFailure(operation, failure)
+    val operationPath = s"$pendingOperationsPath/${packageCoordinate.as[String]}"
+    val failurePath = s"$failedOperationsPath/${packageCoordinate.as[String]}"
+
+    val transaction =
+      client.inTransaction()
+        .delete().forPath(operationPath)
+      .and()
+        .create().forPath(failurePath, Envelope.encodeData(operationFailure))
```

Does this mean that if a package coordinate has a failure associated with it, and then another operation on it fails, that second failure is lost?